### PR TITLE
add metric for cloudcore monitoring

### PR DESCRIPTION
**Pull request by wackxu**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

add metric for cloudcore monitoring

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/apis/componentconfig/cloudcore/v1alpha1/validation/validation.go [hunk 1]
```diff
@@ -43,6 +44,25 @@ func ValidateCloudCoreConfiguration(c *v1alpha1.CloudCoreConfig) field.ErrorList
 	return allErrs
 }
 
+func ValidateCommonConfig(c v1alpha1.CommonConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+	validPort := utilvalidation.IsValidPortNum(int(c.MetricServer.Port))
+	validAddress := utilvalidation.IsValidIP(c.MetricServer.Address)
+
+	if len(validPort) > 0 {
```
Comment (gy95) [inline pkg/apis/componentconfig/cloudcore/v1alpha1/validation/validation.go:52]: no need to check length? And how about adding CommonConfig field `TunnelPort` validation too?

Comment (gy95) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/validation/validation.go:52]: oh, tunnelPort needs being negotiated later.

Comment (wackxu) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/validation/validation.go]: > no need to check length? And how about adding CommonConfig field `TunnelPort` validation too?

The variable names are confusing, rename to validPortErr

File: pkg/apis/componentconfig/cloudcore/v1alpha1/default.go [hunk 2]
```diff
@@ -37,6 +37,10 @@ func NewDefaultCloudCoreConfig() *CloudCoreConfig {
 		},
 		CommonConfig: &CommonConfig{
 			TunnelPort: constants.ServerPort,
+			MetricServer: MetricServer{
+				Address: "0.0.0.0",
+				Port:    9091,
+			},
```
[the remark this answers is not in the document]
Comment (wackxu) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/default.go:42]: rename MetricServer  to MonitorServer and change the default value to 127.0.0.1

File: cloud/pkg/common/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog/v2"
+)
+
+const (
+	metricNamespace = "KubeEdge"
```
[the remark this answers is not in the document]
Comment (wackxu) [reply cloud/pkg/common/metrics/metrics.go]: the namespace is not the kuberntes namespaces and it is just a string that helps to build the final metric name
from the prometheus comment:
```
	// Namespace, Subsystem, and Name are components of the fully-qualified
	// name of the Metric (created by joining these components with
	// "_"). Only Name is mandatory, the others merely help structuring the
	// name. Note that the fully-qualified name of the metric must be a
	// valid Prometheus metric name.
```

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 4]
```diff
@@ -210,24 +216,34 @@ func (uc *UpstreamController) dispatchMessage() {
...
 			uc.ruleStatusChan <- msg
 
 		default:
+			hasError = true
 			klog.Errorf("message: %s, resource type: %s unsupported", msg.GetID(), resourceType)
 		}
+
+		if !hasError {
+			metrics.InflightMessageNum.WithLabelValues(metricType).Inc()
+		}
```
[the remark this answers is not in the document]
Comment (wackxu) [reply cloud/pkg/edgecontroller/controller/upstream.go]: done

image

File: cloud/cmd/cloudcore/app/server.go [hunk 5]
```diff
@@ -73,6 +74,9 @@ kubernetes controller which manages devices so that the device metadata/status d
 				klog.Exit(err)
 			}
 
+			// start monitor server
+			go monitor.ServeMonitor(config.CommonConfig.MonitorServer)
```
[the remark this answers is not in the document]
Comment (wackxu) [reply cloud/cmd/cloudcore/app/server.go:93]: done, use beehivecontext.GetContext in the server

**Comment by wackxu** [comment]
sorry for the deplay, all the comments are resolved,  PTAL @zhu733756 @fisherxu

File: cloud/pkg/common/monitor/monitor.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		)
+	})
+}
+
+func InstallHandlerForPProf(mux *http.ServeMux) {
+	mux.HandleFunc("/debug/pprof/", pprof.Index)
+	mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
+	mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+	mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+	mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
```
Comment (zhu733756) [inline cloud/pkg/common/monitor/monitor.go:68]: do we need /debug/pprof/heap ?

Comment (wackxu) [reply cloud/pkg/common/monitor/monitor.go:68]: It already exits in the `/debug/pprof/`

Comment (zhu733756) [reply cloud/pkg/common/monitor/monitor.go:68]: It works if we don't add sth like "mux.HandleFunc("/debug/pprof/heap", pprof.Heap)"?

Comment (wackxu) [reply cloud/pkg/common/monitor/monitor.go:68]: yes, 

```
root@kubeedge-dev-linux-0004:~/gowork/src/github.com/kubeedge/kubeedge# go tool pprof http://127.0.0.1:9091/debug/pprof/heap
Fetching profile over HTTP from http://127.0.0.1:9091/debug/pprof/heap
Saved profile in /root/pprof/pprof.cloudcore.alloc_objects.alloc_space.inuse_objects.inuse_space.013.pb.gz
File: cloudcore
Build ID: 6aa303827980d57ac12a616d08d9cb9d746d4a73
Type: inuse_space
Time: Nov 9, 2022 at 3:10pm (CST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 9119.46kB, 74.80% of 12192.26kB total
Showing top 10 nodes out of 105
      flat  flat%   sum%        cum   cum%
 1815.09kB 14.89% 14.89%  1815.09kB 14.89%  github.com/kubeedge/kubeedge/cloud/pkg/edgecontroller/controller.NewUpstreamController
 1033.55kB  8.48% 23.36%  1033.55kB  8.48%  regexp.onePassCopy
    1027kB  8.42% 31.79%  1539.01kB 12.62%  k8s.io/apimachinery/pkg/runtime.(*Scheme).AddKnownTypeWithName
 1025.94kB  8.41% 40.20%  1025.94kB  8.41%  regexp/syntax.(*compiler).inst (inline)
    1025kB  8.41% 48.61%     1025kB  8.41%  runtime.allocm
 1024.41kB  8.40% 57.01%  1024.41kB  8.40%  runtime.malg
  605.03kB  4.96% 61.97%   605.03kB  4.96%  github.com/kubeedge/beehive/pkg/core/channel.(*Context).newChannel
  536.37kB  4.40% 66.37%   536.37kB  4.40%  k8s.io/apimachinery/pkg/conversion.ConversionFuncs.AddUntyped
  514.38kB  4.22% 70.59%  1026.94kB  8.42%  k8s.io/api/core/v1.(*ConfigMapList).Unmarshal
  512.69kB  4.21% 74.80%  1025.31kB  8.41%  github.com/emicklei/go-restful.(*WebService).Route

```

Comment (zhu733756) [reply cloud/pkg/common/monitor/monitor.go:68]: OK, thanks your reply

**Comment by zhu733756** [comment]
LGTM

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**